### ISSUE TITLE
Expose via SVN different branches of a GitLab repo?

### ISSUE THREAD

**Issue by db4** [bozaro/git-as-svn#188]
I know git-as-svn does not support (yet) multiple SVN branches layout; [this issue](https://github.com/bozaro/git-as-svn/issues/69) recommends
>   create many svn-repositories for one git-repositry (one repo per branch)

as a workaround. Is it possible for GitLab repos? If yes, can you share an example config doing that?

**Comment by slonopotamus** [comment]
No, it is currently impossible to expose multiple branches of a single repo when using `!gitlabMapping`.

**Comment by db4** [comment]
Can you recommend any workaround? I would shutdown a remaining svn repo and move to git completely  but  svn users need to access multiple branches.

**Comment by slonopotamus** [comment]
Multiple branches per repo are currently only supported for `repositoryMapping: !listMapping`.

However, `listMapping`
1. Doesn't support auto-discovery of newly-created repositories/branches and requires manual config editing and git-as-svn restart when repository list changes.
2. Doesn't support checking user access rights via GitLab, you'll need to specify ACL directly in git-as-svn config. See `repositoryMapping` section in https://github.com/bozaro/git-as-svn/blob/master/cfg/config-local.example

**Comment by db4** [comment]
I've finally found some time to try this approach out. Unfortunately it's not going to work with GitLab-hosted repos. Here is a my config:

```
repositoryMapping: !listMapping
  repositories:
    test-master:
      access: !acl
        groups:
          - name: developers
            users:
              - test
        access:
          - path: /
            allowed:
              - '@developers'
      repository: !git
        pusher: !pushEmbedded
          update: update
          preReceive: pre-receive
          postReceive: post-receive
        createMode: EXAMPLE
        branch: master
        path: /var/opt/gitlab/git-data/repositories/test/test.git
        renameDetection: true

    test-branch:
      access: !acl
        groups:
          - name: developers
            users:
              - test
        access:
          - path: /
            allowed:
              - '@developers'
      repository: !git
        pusher: !pushEmbedded
          update: update
          preReceive: pre-receive
          postReceive: post-receive
        createMode: EXAMPLE
        branch: branches/branch
        path: /var/opt/gitlab/git-data/repositories/test/test.git
        renameDetection: true

userDB: !gitlabUsers {}
```

This way I was able to checkout the repo via SVN, but when I try commit something it fails with
```
Transmitting file data .done
Committing transaction...
svn: E165001: Commit failed (details follow):
svn: E165001: svn: E165001: Hook /var/opt/gitlab/git-data/repositories/test/test.git/hooks/pre-receive failed with output:
GitLab: The project you were looking for could not be found.
```

Probably, there is something special in GitLab hooks (some env vars should be set?). Any thoughts on how to make it work?

**Comment by slonopotamus** [comment]
Aaargh, this is all so weird. GitLab became stupid around 10.x and stopped being able to understand what repo commits are being pushed into (even though hook actually has this information via `GIT_DIR` environment variable and working directory). So GitLab now needs an environment variable with repo id for that (issue #160). But we only do that for `repositoryMapping: !gitlabMapping` because otherwise it isn't clear where we get repository id from :(

> createMode: EXAMPLE

Be careful with this, it wipes stuff.

**Comment by db4** [comment]
> Aaargh, this is all so weird. GitLab became stupid around 10.x (issue #160) and stopped being able to understand what repo commits are being pushed into (even though hook actually has this information via `GIT_DIR` environment variable and working directory). So GitLab now needs an environment variable with repo id for that. But we only do that for `repositoryMapping: !gitlabMapping` because otherwise it isn't clear where we get repository id from :(

Thanks for clarifying.  Then maybe to add `env` node to `pusher` (or `repository`) config that would allow to set arbitrary environment variables before invoking git? One could set `GIT_DIR` manually then. I think it should be relatively easy to implement.

> > createMode: EXAMPLE
> 
> Be careful with this, it wipes stuff.

Thanks. I just took this from the example config.

**Comment by slonopotamus** [comment]
No, that will not work. Besides repo id GitLab needs user id (because they need to know WHO is pushing to repo and whether he has rights to do so). Uh. I guess we need to go back to `!gitlabMapping` path and figure out how to handle branches. We'd need a webhook to handle branch creation/deletion events from gitlab, some kind of filtering on what branches to expose, iterate over available branches on startup (because things could change in repo while git-as-svn is offline), etc.

**Comment by slonopotamus** [comment]
This was fixed in #272 and released in git-as-svn 1.18.0.
